TGUI Engineering... and a bunch of other random things

ShadowLarkens opened the pull request:
This PR adds way too many things to list. It includes changes from these PRs (and probably a few I missed):

https://github.com/VOREStation/VOREStation/pull/7598
https://github.com/VOREStation/VOREStation/pull/8444
https://github.com/VOREStation/VOREStation/pull/8629
https://github.com/VOREStation/VOREStation/pull/8695
https://github.com/VOREStation/VOREStation/pull/8707
https://github.com/VOREStation/VOREStation/pull/8846
https://github.com/VOREStation/VOREStation/pull/8998
https://github.com/VOREStation/VOREStation/pull/9035

I'm not even going to pretend like I can fit all of the interfaces this changes into the body of this PR.
Look at the above PRs for a general idea.

Highlights:
:cl:
add - Added Rapid Pipe Dispenser (RPD).
add - Basically all of engineering and atmos computers converted to TGUI.
add- Vending machines converted to TGUI.
add - PDAs converted to TGUI.
add - Modular Computers Total Conversion
Every AI subsystem left, including the law manager.
add - Wires (w/ colorblindness support).
add - Stack construction menu.
add - pAIs converted.
/:cl:

Atermonera commented on code/modules/tgui/modules/ntos-only/email.dm:
```diff
@@ -0,0 +1,476 @@
+	if(allmails.len > last_message_count)
+		. = 2
+	else if(allmails.len > read_message_count)
+		. = 1
```
These should probably have #defines

Atermonera commented on code/modules/tgui/modules/ntos-only/email.dm:
```diff
@@ -0,0 +1,476 @@
+	switch(action)
+		if("login")
+			log_in()
+			return 1
```
Are... these supposed to be TRUE's?

Atermonera commented on code/modules/tgui/modules/ntos-only/uav.dm:
```diff
@@ -0,0 +1,241 @@
+	// Following roughly the ntnet signal levels
+	// 0 is none
+	// 1 is weak
+	// 2 is strong
```
These should _definitely_ be #defines

Atermonera reviewed (approved):
Well that only took an eternity. I have a few things that I want to specifically test (namely, things that others've asked me to report on how the changes function, like pAIs) but generally speaking the code looks fine. There's a few bits where the encoding of characters like the degrees symbol or that macro y character got changed.

And of course, because I figured it'd happen if other stuff got merged because holy fuck this touches so many files, getting other people to merge other stuff has caused conflicts (And that is part of the reason why I have looked at exactly 0 things outside of determining if they needed to be discussed at staff meetings since starting on this)

ShadowLarkens commented on code/__defines/misc.dm:
```diff
@@ -387,3 +387,86 @@ GLOBAL_LIST_INIT(all_volume_channels, list(
+		"darkolivegreen"= "dark green",	\
+		"darkslateblue" = "dark blue",	\
+		"darkkhaki" 	= "khaki",		\
+		"darkseagreen" 	= "light green",\
```
squint?

Atermonera replied:
Squinting at the mismatch of color names because internet colors are a horrible amalgam of at least two different standards that do not mesh in the slightest because ~compatibility~ _\~quality code!\~_

ShadowLarkens commented on code/datums/wires/robot.dm:
```diff
@@ -1,76 +1,63 @@
+				R.connect_to_ai(select_active_ai(R))
 
-		if (BORG_WIRE_CAMERA)
+		if(WIRE_BORG_CAMERA)
```
not sure

ShadowLarkens commented on code/_helpers/text.dm:
```diff
@@ -21,6 +21,27 @@
+	t = copytext(t,1,limit)
+	for(var/char in strip_chars)
+		var/index = findtext(t, char)
+		while(index)
```
this is a bit over my head to optimize, it seems to work well

ShadowLarkens commented on code/datums/datacore.dm:
```diff
@@ -275,7 +275,6 @@ var/global/list/PDA_Manifest = list()
 		list("cat" = "Medical", "elems" = med),
 		list("cat" = "Science", "elems" = sci),
 		list("cat" = "Cargo", "elems" = car),
-		list("cat" = "Exploration", "elems" = pla), // VOREStation Edit
```
do you have an exploration department?

Atermonera replied:
It's not really a real department, per se, but they are considered separate from all other departments.

ShadowLarkens commented on code/ATMOSPHERICS/components/omni_devices/filter.dm:
```diff
@@ -160,34 +151,41 @@
+			if(!configuring || use_power)
+				return
+			var/new_flow_rate = input(usr,"Enter new flow rate limit (0-[max_flow_rate]L/s)","Flow Rate Control",set_flow_rate) as num
+			set_flow_rate = between(0, new_flow_rate, max_flow_rate)
```
nope

Atermonera commented:
Things I've found while testing (Checked out your branch on my local, running 513.1532):
* Atmos alarm doesn't seem to care about remote control, the control consoles can alter settings just fine when it's set to auto and there are no alarms (The current prerequisite to enable remote control on auto). Not really sure that's a bad thing, but may be worth discussion.
* Employee records console (The stationary laptop thing) doesn't scroll with the scroll wheel. Not an issue for other menus or consoles.
* (Feature request) Pipe dispenser feels a little clunkier having to click back and forth between pipe and device menus, would it be possible to make that a "select all to display" multiple-choice checkbox?
* Gas cooler/heaters: Power level slider is a little confusing at first, that it's specifically a _vertical_ mouse-drag. You appear to be able to click on it and manually submit a number, but doing so does not cause the power level to update.
* Gas canisters: Might be a per-machine issue, but it seems especially affected by the mouse-drag, because I have to move my mouse a significant distance on a 2560x1440 monitor to go from max to min, and it seems like it might be hard to get specific values.
* Gas pumps: Would it be possible to make those a slider, like the heater/cooler temps?
* pAI software interface verb gives ![imes](https://puu.sh/GGbsn/2c9b72c934.png).
* Genuinely curious: Do you know if these are being highlighted differently just by their base color, or is there perchance something else causing it? ![image](https://puu.sh/GGbpO/490f88ccc2.png)

ShadowLarkens commented:
@Atermonera Took me a while to get to this, but...

> Atmos alarm doesn't seem to care about remote control, the control consoles can alter settings just fine when it's set to auto and there are no alarms (The current prerequisite to enable remote control on auto). Not really sure that's a bad thing, but may be worth discussion.

Can we just call this good? I really don't want to dig back into the alarm code to figure out how this broke.

> Employee records console (The stationary laptop thing) doesn't scroll with the scroll wheel. Not an issue for other menus or consoles.

Not easily fixable, it's due to the way element focus in TGUI works. If you click on a section, so a darkened area of the records console screen, it stops scrolling. You have to click specifically on the lighter margins.

> (Feature request) Pipe dispenser feels a little clunkier having to click back and forth between pipe and device menus, would it be possible to make that a "select all to display" multiple-choice checkbox?

Future improvement? :p

> Gas cooler/heaters: Power level slider is a little confusing at first, that it's specifically a vertical mouse-drag. You appear to be able to click on it and manually submit a number, but doing so does not cause the power level to update.

Vertical mouse drag is just how that component works, you get used to it pretty quickly. The manual input not working is a bug, will fix.

> Gas canisters: Might be a per-machine issue, but it seems especially affected by the mouse-drag, because I have to move my mouse a significant distance on a 2560x1440 monitor to go from max to min, and it seems like it might be hard to get specific values.

It's deliberately tuned down compared to other UIs, as the default drag speed makes it way too finnicky to use. Not sure what to do about that one.

> Gas pumps: Would it be possible to make those a slider, like the heater/cooler temps?

Future improvement? :p

> pAI software interface 

Will fix.

> Genuinely curious: Do you know if these are being highlighted differently just by their base color, or is there perchance something else causing it?

I'm pretty sure it's just their base color causing IE to decide to highlight them in a different color.

ShadowLarkens commented:
> Vertical mouse drag is just how that component works, you get used to it pretty quickly. The manual input not working is a bug, will fix.

Fixed.

> > pAI software interface
> Will fix.

Couldn't reproduce the specific bug seen in the screenshot, but fixed other bugs and added a preventative measure that'll hopefully keep the one in the screenshot from happening again.

Atermonera commented on code/ATMOSPHERICS/components/unary/cold_sink.dm:
```diff
@@ -100,7 +100,8 @@
 		if("setPower") //setting power to 0 is redundant anyways
-			var/new_setting = between(0, text2num(params["value"]), 100)
+			var/new_setting = CLAMP(text2num(params["value"]), 0, 100)
+			to_chat(world, "param value was [params["value"]], new_setting was [new_setting]") // DEBUG
```
to_world()
